[FIX] tools: added Image validation in html2plaintext

kmod-odoo opened the pull request:
**Steps to reproduce:**
- Open "Configure Document Layout" in settings.
- Add only an image in the company_details section.

**Issue:**

- The customer added only an image to company_details in their production version 15.2. However, after migrating to version 17, the image appears in company_details but does not show up in the document preview. In version 15.2, there were some changes to company details that caused this issue https://github.com/odoo/odoo/pull/122438. 
- Due to these changes, company_details is processed with the html2plaintext() function, which removes all the tags from company_details and only links and text will remain.
- This function does not recognize the image element and removes it along with other elements. I have prepared a fix that recognizes the image element from company_details and prevents it from getting discarded.
- Video reference from RunBot version 17: https://drive.google.com/file/d/1Ml62xIV1mhwKwU8KRHdc-qfTFCZhEqY0/view?usp=sharing
- I have also added a test case for html2plaintext function in test_html2plaintext function by backporting this method from 
master.

OPW : [3945490](https://www.odoo.com/odoo/my-tasks/3945490?cids=2)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/169027.png)](https://mergebot.odoo.com/odoo/odoo/pull/169027)

kmod-odoo commented:
@odoo/rd-security Can you please review it as the customer is waiting for this fix for their production migration?

mart-e commented:
@kmod-odoo security-related not much to say but maybe modify the `test_html2plaintext` to include your new usecase.
Also many images don’t have an `alt` so would be good to handle it properly (don’t add an empty space)

kmod-odoo commented:
@mart-e  If I add another function, test_html2plaintext, for my use case, it will contain a lot of duplicate code from html2plaintext. Should I proceed with adding it? Additionally, I have handled the alt attribute by adding the alt text instead of leaving an empty space.

mart-e commented:
> If I add another function, test_html2plaintext, for my use case

No I mean the existing `test_html2plaintext` method. To cover your usecase.

> Additionally, I have handled the alt attribute by adding the alt text instead of leaving an empty space.

You litterally add the word `alt` if there is no alt-text. No I mean don’t add alt text and do not add empty spaces if there is none.

Where does this task comes from? You mention it’s a fix but do you have a support ticket with it or something?

Should probably be reviewed by the framework team and could be in master instead if no reason for stable.

kmod-odoo commented:
Sorry for the misunderstanding. I have now added the name of the image instead of the alt attribute, so there is no need to worry about whether alt is present or not.I have also added a task reference in the PR description. Regarding the test_html2plaintext function, it is present in the master branch. However, my PR is currently targeting version 16.0, where that function does not exist. What is your suggestion on this? Should I target the master branch instead?

mart-e commented:
> What is your suggestion on this? Should I target the master branch instead?

I am suggesting that you should check with your team leader/guru/framework team which one is best as this is not really security related and all is fine with me :)

kmod-odoo commented:
> I am suggesting that you should check with your team leader/guru/framework team which one is best as this is not really security related and all is fine with me :)

@mart-e  Thank you for your feedback. @epictete, could you please review this PR? It is related to the PR you approved earlier: https://github.com/odoo/odoo/pull/122438.

malb-odoo commented on odoo/tools/mail.py:
```diff
@@ -380,6 +380,14 @@ def html2plaintext(html, body_id=None, encoding='utf-8'):
+        src = img.get('src')
+        if src:
+            i += 1
+            img.tag = 'span'
+            img.text = '%s [%s]' % (src.split('/')[-1].split('.')[0], i)
```
I know that for the url we have that too, but i think that the tag and the text are useless for the image at least so it can be removed :smile: Moreover if we really wanted to keep the text, in my opinion it's not robust enough specially with no test, if for any reason the src is not formatted like this it will put something else has the text for the img :smile:

kmod-odoo replied:
> I know that for the url we have that too, but i think that the tag and the text are useless for the image at least so it can be removed.

No, we can't remove that because we need to add some text in place of the img tag so that it remains after all tags are removed. The html2plaintext function should return something instead of an empty string in cases where company_details only contain image tags.
https://github.com/odoo/odoo/blob/4a13bbc1b9d5f449e51c065fd9530a099d4cfec4/odoo/tools/mail.py#L396

epictete replied:
As @malb-odoo mentioned, splitting `src` using `/` and `.`, and hoping (because that's all it is) to retrieve the image name is really not a good idea :confused: 
Try using drag and drop to insert the image in the `company_details` field for example :wink: 
(Please update your test to include that use case)
It shouldn't be to hard to do that with a regex :nerd_face:

malb-odoo reviewed (commented):
Thanks for this fix :smile: Small remark and also the commit message is a bit wrong for the number of the ticket follow this convention: https://www.odoo.com/documentation/17.0/contributing/development/git_guidelines.html#commit-message-structure

kmod-odoo commented:
Hello @malb-odoo, PR is ready for review; could you please review it?

malb-odoo reviewed (approved):
Thanks for the addition of a test ! LGTM 
cc: @epictete

epictete reviewed (approved):
Thanks for the quick update, looks better indeed :+1: 

@robodoo r+
